### PR TITLE
part 3

### PR DESCRIPTION
**Pull request by Alexdelia**

### CODE REVIEW

File: src/train/src/learn.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+			for (model_label, thetas) in &mut model.weights {
+				*thetas = guess(
+					thetas,
+					&datasets.training,
+					arg.learning_rate,
+					label == model_label,
+				);
+			}
+		}
+	}
```
Comment (justincapik) [inline src/train/src/learn.rs]: update thetas after one iteration, only add them together at end of iteration of at the end of batch

File: src/train/src/learn.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+fn guess(thetas: &[Float], rows: &[Vec<Float>], learning_rate: Float, truth: bool) -> Vec<Float> {
+	let mut new_thetas = thetas.to_vec();
+	let truth = if truth { 1.0 } else { 0.0 };
+
+	for (i, new_theta) in new_thetas.iter_mut().enumerate() {
+		let sum = rows
+			.iter()
+			.map(|row| (hypothesis(row, thetas) - truth) * row[i])
```
Comment (justincapik) [inline src/train/src/learn.rs]: can do different loss functions for bonus:
par exemple, Squared Error:
```suggestion
			.map(|row| ((hypothesis(row, thetas))^2 - truth^2) * row[i])
```
or Huber loss

**Review by justincapik — COMMENTED** [review]
beautiful

**Comment by justincapik** [comment]
incredible 👏👏👏
